packages/starlight/utils/plugins.ts

```diff
@@ -0,0 +1,143 @@
+import type { AstroIntegration, AstroIntegrationLogger } from 'astro';
+import { z } from 'astro/zod';
+import type { StarlightUserConfig } from '../utils/user-config';
+import { errorMap } from '../utils/error-map';
+
+/**
+ * Runs Starlight plugins in the order that they are configured after validating the plugins config
+ * and returns the final user config that may have been updated by the plugins and a list of any
+ * integrations added by the plugins.
+ * Note that the returned config is not validated and should still be validated by the caller.
+ */
+export async function runPlugins(
+	{ plugins, ...config }: StarlightUserConfigWithPlugins,
+	logger: AstroIntegrationLogger
+) {
+	const parsedPluginsConfig = starlightPluginsConfigSchema.safeParse({ plugins }, { errorMap });
+
+	if (!parsedPluginsConfig.success) {
+		throw new Error(
+			'Invalid plugins config passed to starlight integration\n' +
+				parsedPluginsConfig.error.issues.map((i) => i.message).join('\n')
+		);
+	}
+
+	let userConfig = config;
+	const integrations: AstroIntegration[] = [];
+
+	for (const { name, plugin } of parsedPluginsConfig.data.plugins) {
+		await plugin({
+			config: userConfig,
+			logger: logger.fork(name),
+			updateConfig(newConfig) {
+				userConfig = { ...userConfig, ...newConfig };
+			},
+			addIntegration(integration) {
+				integrations.push(integration);
+			},
+		});
+	}
+
+	return { integrations, userConfig };
+}
+
+// https://github.com/withastro/astro/blob/144815c6ff295fe54eef3c641ad929e3c7b2b2fd/packages/astro/src/core/config/schema.ts#L101
+const astroIntegrationSchema = z.object({
+	name: z.string(),
+	hooks: z.object({}).passthrough().default({}),
+}) as z.Schema<AstroIntegration>;
+
+/**
+ * A plugin `config` and `updateConfig` argument are purposely not validated using the Starlight
+ * user config schema but properly typed for user convenience because we do not want to run any of
+ * the Zod `transform`s used in the user config schema when running plugins.
+ * The final user config should be validated by the caller after running all the plugins.
+ */
+const starlightPluginSchema = z.object({
```

Comment:
I’m not sure I follow why we can’t include `plugins` in the main user config schema? Would you mind explaining a bit more?

Reply:
Is this question related to the highlighted comment or more of a general question?

If this is directly relateed to the comment, I think my comment is not obvious enough. I try to explain why we are not using something like:

```js
updateConfig: z.function(
  z.tuple([UserConfigSchema]),
  //          ^ this is the problem described in the comment
  z.void()
)
```

If this is not directly to the comment, this would be technically possible to add the plugins to the user config schema. The reason I did not do that are:

- There is __at the moment__ in the code only 1 place (the Starlight integration input) where we need a type that contains a config with the plugins. 
- Past that point, we only need the config without the plugins (when passing it down to plugins, when a plugin updates the config, for passing it to the Vite plugin, etc.)
- I did not want to re-validate the plugins config when re-validating the user config after the plugins have run.
- I thought it would be easier to use a TypeScript union once vs omitting the plugin from the user config schema with the Zod API to get the same result.

This could definitely be revisited if you think it would be better to have the plugins in the user config schema.

Reply:
> Past that point, we only need the config without the plugins (when passing it down to plugins, when a plugin updates the config, for passing it to the Vite plugin, etc.)

Wouldn’t other plugins want to know what plugins they are configured alongside potentially?

Makes sense for the Vite plugin though, where functions would be stripped during serialization.

Reply:
> Wouldn’t other plugins want to know what plugins they are configured alongside potentially?

I didn't think of this but it's a very good point.

So I guess a potential solution would be:

- Add `plugins` to the user config schema and refactor the various types/calls to follow this change
- Pass down the `plugins` too to the provided `config` to a plugin
- As per your comment above, check if `plugins` is being mutated in the `updateConfig()` hook and error if so

Is this something that looks good to you?

Reply:
Yeah, sounds good to me!